File: leetcode/986_interval_list_intersections.py
"""
T: O(A + B)
S: O(C), where C = min(A, B)

Walk through both of the lists and check if two current intervals intersect.
If so - add them to the result. Move the pointer in the list which interval
ends earlier.
"""

from typing import List


class Solution:

    def intervalIntersection(self, A: List[List[int]],
                             B: List[List[int]]) -> List[List[int]]:
        result = []  # type: List[List[int]]
        if not A or not B:
            return result
        i = j = 0
        a, b = len(A), len(B)
        while i < a and j < b:
            intA = A[i]
            intB = B[j]
            start = max(intA[0], intB[0])
            end = min(intA[1], intB[1])
            if start <= end:
                result.append([start, end])
            if intA[1] < intB[1]:
                i += 1
            else:
                j += 1
        return result

File: leetcode/java/986IntervalListIntersections.java
/**
 * T: O(A+B) S: O(A+B)
 *
 * <p>To check if there is an intersection we check if the bigger of starts is smaller than the
 * smaller of ends. If so - we found an intersection and add it to the result. In any case, we
 * advance the smaller interval. In the end we just return found intersetions.
 */
import java.util.ArrayList;
import java.util.List;

class Solution {
  public int[][] intervalIntersection(int[][] A, int[][] B) {
    List<int[]> intersections = new ArrayList();
    int a = 0;
    int b = 0;

    while (a < A.length && b < B.length) {
      int start = Math.max(A[a][0], B[b][0]);
      int end = Math.min(A[a][1], B[b][1]);
      if (start <= end) intersections.add(new int[] {start, end});

      if (A[a][1] < B[b][1]) a++;
      else b++;
    }

    return intersections.toArray(new int[intersections.size()][]);
  }
}
